warehouse/models/staging/payments/littlepay/stg_littlepay__product_data.sql

```diff
@@ -29,12 +29,79 @@ stg_littlepay__product_data AS (
         {{ safe_cast('capping_overlap', 'TIME') }} AS capping_overlap,
         {{ trim_make_empty_string_null('capping_application_level') }} AS capping_application_level,
-        _line_number,
+        CAST(_line_number AS INTEGER) AS _line_number,
         `instance`,
         extract_filename,
+        {{ extract_littlepay_filename_ts() }} AS littlepay_export_ts,
+        {{ extract_littlepay_filename_date() }} AS littlepay_export_date,
         ts,
+        -- hash all content not generated by us to enable deduping full dup rows
+        -- hashing at this step will preserve distinction between nulls and empty strings in case that is meaningful upstream
+        {{ dbt_utils.generate_surrogate_key(['participant_id', 'product_id', 'product_code',
+            'product_description', 'product_type', 'activation_type', 'product_status',
+            'created_date', 'capping_type', 'multi_operator', 'capping_start_time',
+            'capping_end_time', 'rules_transaction_types', 'rules_default_limit',
+            'rules_max_fare_value', 'scheduled_start_date_time', 'scheduled_end_date_time',
+            'all_day', 'weekly_cap_start_day', 'number_of_days_in_cap_window',
+            'capping_duration', 'number_of_transfer', 'capping_time_zone', 'capping_overlap',
+            'capping_application_level']) }} AS _content_hash,
     FROM source
-    QUALIFY ROW_NUMBER() OVER (PARTITION BY product_id ORDER BY littlepay_export_ts DESC) = 1
+),
+
+add_keys_drop_full_dupes AS (
+    SELECT
+        *,
+        -- generate keys now that input columns have been trimmed & cast and files deduped
+        {{ dbt_utils.generate_surrogate_key(['littlepay_export_ts', '_line_number', 'instance']) }} AS _key,
+        product_id AS _payments_key,
+    FROM clean_columns
+    {{ qualify_dedupe_full_duplicate_lp_rows() }}
+),
+
+stg_littlepay__product_data AS (
+    SELECT
+        participant_id,
+        product_id,
+        product_code,
+        product_description,
+        product_type,
+        activation_type,
+        product_status,
+        created_date,
+        capping_type,
+        multi_operator,
+        capping_start_time,
+        capping_end_time,
+        rules_transaction_types,
+        rules_default_limit,
+        rules_max_fare_value,
+        scheduled_start_date_time,
+        scheduled_end_date_time,
+        all_day,
+        weekly_cap_start_day,
+        number_of_days_in_cap_window,
+        capping_duration,
+        number_of_transfer,
+        capping_time_zone,
+        capping_overlap,
+        capping_application_level,
+        _line_number,
+        `instance`,
+        extract_filename,
+        littlepay_export_ts,
+        littlepay_export_date,
+        ts,
+        _key,
+        _payments_key,
+        _content_hash,
+    FROM add_keys_drop_full_dupes
+    -- Some products change in form over time, e.g. getting different 'capping_type' values or
```

Comment:
Similar question to above.... I am wondering whether these need to be handled as type 2 or if we should treat these as corrections. 🤔 Because the question would be if we're looking at older data, should we access the capping conditions that were in effect at the time? Or the capping conditions as they are now?